deepinv/models/aliasfree.py

```diff
@@ -0,0 +1,740 @@
+                inp_norm = torch.norm(inp, p=2, dim=(-1, -2), keepdim=True)
+            out = self.filt(inp)
+            if self.scale_l2:
+                out_norm = torch.norm(out, p=2, dim=(-1, -2), keepdim=True)
+                out = out * (inp_norm / (out_norm + self.eps))
+            return out[:, :, :: self.stride, :: self.stride]
+
+        elif self.filt_size == 1:
+            if self.pad_off == 0:
+                return inp[:, :, :: self.stride, :: self.stride]
+            else:
+                return self.pad(inp)[:, :, :: self.stride, :: self.stride]
+
+        else:
+            return self.filt(inp)[:, :, :: self.stride, :: self.stride]
+
+    @staticmethod
+    def get_rect(filt_size):
+        if filt_size == 1:
+            a = np.array(
+                [
+                    1.0,
+                ]
+            )
+        elif filt_size == 2:
+            a = np.array([1.0, 1.0])
+        elif filt_size == 3:
+            a = np.array([1.0, 2.0, 1.0])
+        elif filt_size == 4:
+            a = np.array([1.0, 3.0, 3.0, 1.0])
+        elif filt_size == 5:
+            a = np.array([1.0, 4.0, 6.0, 4.0, 1.0])
+        elif filt_size == 6:
+            a = np.array([1.0, 5.0, 10.0, 10.0, 5.0, 1.0])
+        elif filt_size == 7:
+            a = np.array([1.0, 6.0, 15.0, 20.0, 15.0, 6.0, 1.0])
+
+        return a
+
+    def __repr__(self):
+        return (
+            f"BlurPool(channels={self.channels}, pad_type={self.pad_type}, "
+            f" stride={self.stride}, filter_type={self.filter_type},  filt_size={self.filt_size}, "
+            f"scale_l2={self.scale_l2})"
+        )
+
+
+class Filter(nn.Module):
+    def __init__(
+        self, filt, channels, pad_type=None, pad_sizes=None, scale_l2=False, eps=1e-6
+    ):
+        super(Filter, self).__init__()
+        self.register_buffer("filt", filt[None, None, :, :].repeat((channels, 1, 1, 1)))
+        if pad_sizes is not None:
+            self.pad = get_pad_layer(pad_type)(pad_sizes)
+        else:
+            self.pad = None
+        self.scale_l2 = scale_l2
+        self.eps = eps
+
+    def forward(self, x):
+        if self.scale_l2:
+            inp_norm = torch.norm(x, p=2, dim=(-1, -2), keepdim=True)
+        if self.pad is not None:
+            x = self.pad(x)
+        out = F.conv2d(x, self.filt, groups=x.shape[1])
+        if self.scale_l2:
+            out_norm = torch.norm(out, p=2, dim=(-1, -2), keepdim=True)
+            out = out * (inp_norm / (out_norm + self.eps))
+        return out
+
+
+def get_pad_layer(pad_type):
+    if pad_type in ["refl", "reflect"]:
+        PadLayer = nn.ReflectionPad2d
+    elif pad_type in ["repl", "replicate"]:
+        PadLayer = nn.ReplicationPad2d
+    elif pad_type == "zero":
+        PadLayer = nn.ZeroPad2d
+    elif pad_type == "circular":
+        PadLayer = circular_pad
+    else:
+        print("Pad type [%s] not recognized" % pad_type)
+    return PadLayer
+
+
+class AliasFreeDenoiser(nn.Module):
```

Comment:
Should this be called AliasFreeUNet? I'm not sure...

Reply:
Good idea! Another option might be EquivariantUNet. What do you think? I'm changing it to your suggestion for now.

Reply:
In hindsight I'd say we care more about the equivariance and removing aliasing is just one of the things we need to do to get it. I've renamed it to EquivariantUNet.